### PR TITLE
Run llama-index examples in CI

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/12794?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12794/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12794
```

</p>
</details>

### What changes are proposed in this pull request?

Add LlamaIndex examples to the CI tests.

Now we can run examples for OpenAI-required flavors, such as openai, langchain. Will add them as follow-up (need some update on examples supposedly)

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: requirements/extra-ml-requirements.txt [hunk 1]
```diff
@@ -66,6 +66,8 @@ tiktoken
 tenacity
 # Required by mlflow.llama_index
 llama_index
+# Required for an agent example of mlflow.llama_index
+llama-index-agent-openai
```
Comment (serena-ruan) [inline requirements/extra-ml-requirements.txt:70]: Can we add this into test requirements?

File: tests/examples/test_examples.py [hunk 2]
```diff
@@ -38,6 +38,15 @@ def clean_up_mlflow_virtual_environments():
             shutil.rmtree(path)
 
 
+@pytest.fixture(scope="module", autouse=True)
+def mock_openai():
+    # Some examples includes OpenAI API calls, so we start a mock server.
+    with start_mock_openai_server() as base_url:
+        os.environ["OPENAI_API_BASE"] = base_url
+        os.environ["OPENAI_API_KEY"] = "test"
+        yield
```
Comment (harupy) [inline tests/examples/test_examples.py:47]:
```suggestion
        with pytest.MonkeyPatch.context() as mp:
            mp.setenv("OPENAI_API_BASE", base_url)
            mp.setenv("OPENAI_API_KEY", "test")
            yield
```

to minimize side effects

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/12794/files#r1699952707 is addressed!

File: requirements/extra-ml-requirements.txt (hunk 1 above)
Comment (B-Step62) [reply requirements/extra-ml-requirements.txt:70]: I think `test-requirements.txt` does not flavor tests e.g. it doesn't contain `langchain`, `transformers`, etc.

Comment (serena-ruan) [reply requirements/extra-ml-requirements.txt:70]: Can we add in ml-package-versions.yml

Comment (B-Step62) [reply requirements/extra-ml-requirements.txt:70]: No, we don't need this for cross-version test. It is only required for example test.

**Pull request merged**